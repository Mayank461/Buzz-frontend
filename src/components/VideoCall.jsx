import React, { useEffect, useRef, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import { socket } from '../App';
import Peer from 'peerjs';

function genRoomID(id, id2) {
  if (id > id2) return `${id}-${id2}`;
  return `${id2}-${id}`;
}

const VideoCall = ({ user }) => {
  const [searchParams] = useSearchParams();

  const [myPeerId, setMyPeerId] = useState('');
  const [remotePeerID, setRemotePeerId] = useState(null);
  const currentUserVideoRef = useRef(null);
  const remoteVideoRef = useRef(null);
  const peerInstance = useRef(null);

  const myStream = useRef(null);
  const remoteStream = useRef(null);

  useEffect(() => {
    const peer = new Peer();
    const recipient = searchParams.get('recipient');
    setRemotePeerId(searchParams.get('rpeerid'));

    peer.on('open', (id) => {
      setMyPeerId(id);

      if (!searchParams.get('rpeerid')) {
        socket.emit('CallNotify', recipient, {
          id: user._id,
          peerid: id,
          picture_url: user.picture_url,
          name: user.firstname + ' ' + user.lastname,
        });
      }
    });

    peer.on('call', (call) => {
      var getUserMedia =
        navigator.getUserMedia ||
        navigator.webkitGetUserMedia ||
        navigator.mozGetUserMedia;

      getUserMedia({ video: true }, (mediaStream) => {
        currentUserVideoRef.current.srcObject = mediaStream;
        currentUserVideoRef.current.play();
        call.answer(mediaStream);
        call.on('stream', function (remoteStream) {
          remoteVideoRef.current.srcObject = remoteStream;
          remoteVideoRef.current.play();
        });
      });
    });

    peerInstance.current = peer;

    call(remotePeerID);

    return unmount;
  }, []);

  function unmount() {
    currentUserVideoRef.current.srcObject.getTracks().forEach((s) => s.stop());
    remoteVideoRef.current.srcObject.getTracks().forEach((s) => s.stop());
  }

  const call = (rPeerId) => {
    var getUserMedia =
      navigator.getUserMedia ||
      navigator.webkitGetUserMedia ||
      navigator.mozGetUserMedia;

    getUserMedia({ video: true }, (mediaStream) => {
      currentUserVideoRef.current.srcObject = mediaStream;
      currentUserVideoRef.current.play();

      const call = peerInstance.current.call(rPeerId, mediaStream);

      call.on('stream', (remoteStream) => {
        remoteVideoRef.current.srcObject = remoteStream;
        remoteVideoRef.current.play();
      });
    });
  };

  return (
    <div className="position-relative">
      <button
        onClick={() => {
          call(remotePeerID);
        }}
      >
        click me
      </button>
      <button
        onClick={() => {
          unmount();
        }}
      >
        off cam
      </button>
      <video ref={currentUserVideoRef} className="myvideoScreen" />
      <video ref={remoteVideoRef} className="remoteVideoScreen" />
    </div>
  );
};

export default VideoCall;
