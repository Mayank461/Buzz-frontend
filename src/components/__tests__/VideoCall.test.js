import VideoCall from '../VideoCall';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

const user = {
  _id: 1,
  firstname: 'Test',
  lastname: 'User',
  friends: { myFriends: [], mySentRequests: [{}] },
};

describe('layout test', () => {
  it('Snapshot match', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <VideoCall user={user} />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
