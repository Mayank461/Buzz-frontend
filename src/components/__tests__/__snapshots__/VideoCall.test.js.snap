// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout test Snapshot match 1`] = `
<div
  className="position-relative parent-vc"
>
  <div
    className="hidden"
    id="fireByAccepter"
    onClick={[Function]}
  />
  <video
    className="myvideoScreen"
    muted={true}
  />
  <video
    className="remoteVideoScreen"
  />
  <div
    className="vcdock d-flex justify-content-center"
  >
    <div
      className="icon-roundbox  "
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="fa fa-video-camera"
      />
    </div>
    <div
      className="icon-roundbox bg-danger"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="fa fa-phone"
      />
    </div>
  </div>
</div>
`;
